File: leetcode/201-300/232-implement-queue-using-stacks/bench/queue_using_stacks.py
"""Benchmark workload for LeetCode #232 — Implement Queue using Stacks (Python; scale lane)."""


class MyQueue:
    def __init__(self):
        self.inbox = []
        self.outbox = []

    def push(self, x):
        self.inbox.append(x)

    def refill(self):
        if not self.outbox:
            while self.inbox:
                self.outbox.append(self.inbox.pop())

    def pop(self):
        self.refill()
        return self.outbox.pop()

    def peek(self):
        self.refill()
        return self.outbox[-1]

    def empty(self):
        return not self.inbox and not self.outbox


def main():
    n = 75000000
    cap = 4096
    mask = 1048575

    q = MyQueue()
    sz = 0
    sink = 0
    state = 12345

    for _ in range(n):
        state = (state * 1103515245 + 12345) & 2147483647
        if q.empty() or (state % 2 == 0 and sz < cap):
            q.push(state & mask)
            sz += 1
        elif state % 4 == 0:
            sink += q.peek()
        else:
            sink += q.pop()
            sz -= 1
    print(sink)


main()

File: leetcode/201-300/232-implement-queue-using-stacks/bench/queue_using_stacks.rs
struct MyQueue {
    inbox: Vec<i64>,
    outbox: Vec<i64>,
}

impl MyQueue {
    fn new() -> MyQueue {
        MyQueue {
            inbox: Vec::new(),
            outbox: Vec::new(),
        }
    }

    fn push(&mut self, x: i64) {
        self.inbox.push(x);
    }

    fn refill(&mut self) {
        if self.outbox.is_empty() {
            while let Some(v) = self.inbox.pop() {
                self.outbox.push(v);
            }
        }
    }

    fn pop(&mut self) -> i64 {
        self.refill();
        self.outbox.pop().unwrap_or(-1)
    }

    fn peek(&mut self) -> i64 {
        self.refill();
        self.outbox[self.outbox.len() - 1]
    }

    fn empty(&self) -> bool {
        self.inbox.is_empty() && self.outbox.is_empty()
    }
}

fn main() {
    let n: i64 = 75000000;
    let cap: i64 = 4096;
    let mask: i64 = 1048575;

    let mut q = MyQueue::new();
    let mut sz: i64 = 0;
    let mut sink: i64 = 0;
    let mut state: i64 = 12345;

    for _ in 0..n {
        state = (state * 1103515245 + 12345) & 2147483647;
        if q.empty() || (state % 2 == 0 && sz < cap) {
            q.push(state & mask);
            sz += 1;
        } else if state % 4 == 0 {
            sink += q.peek();
        } else {
            sink += q.pop();
            sz -= 1;
        }
    }
    println!("{}", sink);
}
